modules/renter/repair.go

```diff
@@ -26,323 +25,374 @@ var renewThreshold = func() types.BlockHeight {
 	case "dev":
 		return 200
 	default:
-		return 144 * 7 * 3 // 3 weeks - to soon be 6 weeks.
+		return 144 * 7 * 6
 	}
 }()
 
-// hostErr and hostErrs are helpers for reporting repair errors. The actual
-// Error implementations aren't that important; we just need to be able to
-// extract the NetAddress of the failed host.
-
-type hostErr struct {
-	host modules.NetAddress
-	err  error
-}
-
-func (he hostErr) Error() string {
-	return fmt.Sprintf("host %v failed: %v", he.host, he.err)
+type chunkGaps struct {
+	contracts []types.FileContractID
+	pieces    []uint64
+	numGaps   int
 }
 
-type hostErrs []*hostErr
-
-func (hs hostErrs) Error() string {
-	var errs []error
-	for _, h := range hs {
-		errs = append(errs, h)
+func (r *Renter) addFileToRepairMatrix(file *file, availableWorkers map[types.FileContractID]struct{}, repairMatrix map[string]*chunkGaps, gapCounts map[int]int) {
+	// Flatten availableWorkers into a list of contracts.
+	contracts := make([]types.FileContractID, 0)
+	for contract := range availableWorkers {
+		contracts = append(contracts, contract)
 	}
-	return build.JoinErrors(errs, "\n").Error()
-}
 
-// repair attempts to repair a file chunk by uploading its pieces to more
-// hosts.
-func (f *file) repair(chunkIndex uint64, missingPieces []uint64, r io.ReaderAt, hosts []contractor.Editor) error {
-	// read chunk data and encode
-	chunk := make([]byte, f.chunkSize())
-	_, err := r.ReadAt(chunk, int64(chunkIndex*f.chunkSize()))
-	if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
-		return err
+	// If the file is not being tracked, don't add it to the repair matrix.
+	_, exists := r.tracking[file.name]
+	if exists {
+		return
 	}
-	pieces, err := f.erasureCode.Encode(chunk)
-	if err != nil {
-		return err
+
+	// For each chunk, create a map from the chunk to the pieces that it
+	// has, and to the contracts that have that chunk.
+	chunkCount := file.numChunks()
+	contractMap := make(map[uint64][]types.FileContractID)
+	pieceMap := make(map[uint64][]uint64)
+	for i := uint64(0); i < chunkCount; i++ {
+		contractMap[i] = make([]types.FileContractID, 0)
+		pieceMap[i] = make([]uint64, 0)
```

Comment:
I generally like being explicit about making things b/c I often `exists` a lot, also for slices I feel like there's a pretty strong difference between an empty slice and `nil`. I don't think I use either of those things in this function, but it still bothers me to have a map that's missing elements when I expect them to be there.